Makefile

```diff
@@ -2363,14 +2383,21 @@ checkout_folly:
 	@# Pin to a particular version for public CI, so that PR authors don't
 	@# need to worry about folly breaking our integration. Update periodically
 	cd third-party/folly && git reset --hard beacd86d63cd71c904632262e6c36f60874d78ba
-	@# A hack to remove boost dependency.
-	@# NOTE: this hack is not needed if using FBCODE compiler config
-	perl -pi -e 's/^(#include <boost)/\/\/$$1/' third-party/folly/folly/functional/Invoke.h
 	@# NOTE: this hack is required for clang in some cases
 	perl -pi -e 's/int rv = syscall/int rv = (int)syscall/' third-party/folly/folly/detail/Futex.cpp
 	@# NOTE: this hack is required for gcc in some cases
 	perl -pi -e 's/(__has_include.<experimental.memory_resource>.)/__cpp_rtti && $$1/' third-party/folly/folly/memory/MemoryResource.h
 
+build_folly:
```

Comment:
In my experience, quality builds are divided into (up to) two distinct phases:
1) download any dependencies (without running the compiler)
2) run the compiler and maybe tests (without hitting the network)

And there should be a "clean" step that gets you back to the state after 1 so that you can rebuild with a different compiler or compilation settings. Ideally, each of those two phases requires only one command invocation.

What that means here (I think):
* We should download, etc. any/all the folly dependencies as part of the checkout_folly step.
* If configured to use folly, we should automatically build libfolly.a if it doesn't exist as part of normal "make check" or "make all". We don't need to do dependency tracking in folly, because we are not intending to develop on it from rocksdb repo.
* If third-party/folly exists, then clean folly build outputs as part of our "make clean", so that libfolly.a is rebuilt automatically if changing compilers, etc.